draft-ietf-quic-multipath.md

```diff
@@ -398,7 +398,11 @@ if any.
 
 The receiver of a PATH_ABANDON frame SHOULD NOT release its resources
 immediately, but SHOULD wait for the reception of the RETIRE_CONNECTION_ID
-frame for the used connection IDs or 3 RTOs.
+frame for the used connection IDs but at least three times the current
+Probe Timeout (PTO) interval as defined in {{Section 6.2. of QUIC-RECOVERY}}.
+This is inline with the requirement of {{Section 10.2 of QUIC-TRANSPORT}}
+to ensure that paths close cleanly and that delayed or reordered packets
+are properly discarded.
 
```

Comment:
This does not read very well: "SHOULD wait for the reception of the RETIRE_CONNECTION_ID
frame for the used connection IDs but at least three times the current
Probe Timeout (PTO) interval as defined in {{Section 6.2. of QUIC-RECOVERY}}." I don't know how to interpret it.

I think what you mean is: "the receiver of a PATH_ABANDON frame SHOULD NOT release its resources
immediately, but SHOULD wait for at least 3 RTOs after the last time a packet was sent using the corresponding CID or until it receives a RETIRE_CONNECTION_ID frame for that CID, whichever happens sooner."

Then there is the separate question of the sender behavior. I think we should says something like: "Nodes that Abandon a PATH SHOULD wait for 3 RTOs before sending a RETIRE_CONNECTION_ID frame for the corresponding CID. However, they MAY issue the RETIRE_CONNECTION_ID frame sooner, for example because of resource constraints or because the peer asked to retire all connection identifier before a specific sequence number. This may cause the peer node to conclude that no acknowledgement will be received for packets sent with that CID."

Reply:
What you propose in your text is exactly what we need to clarify. I think based on {{Section 6.2. of QUIC-RECOVERY}} you always need to wait for at least 3 PTO. If we say it that way, I don't think there is a need to add the second part you propose to wait with the sending of RETIRE_CONNECTION_ID frame . That would be another solution but more complex.

Reply:
Yes, agree. I'll give it a try. But we do agree that's the right thing?

Reply:
I think we agree on the meaning, but we should indeed improve the presentation. What about replacing the first (very) long sentence by this?

```
Upon the reception of a PATH_ABANDON, the endpoint SHOULD delay the release
of the associated path resource until either a) the peer has retired the
Connection ID it used on the abandoned path, or b) at least three times the
current Probe Timeout (PTO) interval as defined in {{Section 6.2. of QUIC-RECOVERY}}.
```

Reply:
Okay this means we do not agree. I believe you always have to wait at least threes PTOs because of re-ordered packets. I guess if you receive the retired CID frame that means that the peer will not send any new packets but you can only be sure that there is no reordering if you received all packets before the packet with the retired CID frame, however, you don't really know how far back you have to look or there could be gaps introduced by the sender. So checking that is not really practical and always waiting for 3 PTOs is therefore just simpler. Or is there is strong reason why it would be needed to release resources on receipt of the retired CID frame already?

I note that this discussion is also related to the question if we change the path ID with the CID or not. If we retire the path ID together with the CID, it makes more sense to act on the retired CID frame because in this case even if you receive reordered packets, you would not be able to process them as you don't have a valid path ID.

Reply:
I agree that endpoint should always wait at least three PTOs before it releases resources of the corresponding path, but the sentence need to be fixed. Maybe we need 2 states "Closing Path State" and "Draining Path State": 
- After the endpoint sends the PATH_ABANDON frame, it enters into the "Closing Path State", and needs to wait for at least three PTOs before releasing the CID resource of the corresponding path. 
- The "Draining Path State" is entered once an endpoint receives a PATH_ABANDON frame, which indicates that its peer is closing or draining on this path. "Draining Path State" last for at least three PTOs, before it ends, the endpoint SHOULD NOT release the corresponding CID resouce, until it receives a RETIRE_CONNECTION_ID frame for that CID.

Reply:
This PR is actually also connects to the question on path  ID.

First of all it is not clear to me what an endpoints need to do if it first receives a retired CID frame and then a (re-order) packet with that CID. In order word is an endpoint supposed to retire/forget the CID immediately when it received the retire CID frame and discard any packets with the old CID or might it also what 1-3 PTOs for re-order packet? I didn't find anything in RFC9000 but maybe somebody else that is better in finding this has a pointer to right part...?

If the former, meaning you remove it immediately and ignore old packets, that would also mean that you don't have a valid path ID anymore and can't process any ACKs or other control frame with the old path ID anymore. As such you also don't have to wait for 3 PTOs as you might receive packet/frames but can process them anymore. In this case (assuming that the path ID is connected to the CID), one solution could be to recommend to wait 3PTOs before you send the retired CID frame (or just accepting that this might happens and you have time retransmit some packets unnecessarily if you change the CID).

If you have a stable path ID, you can retire immediately and ignore packets with the old CID but still process any control frames that are received in packets with the new CID, and respectively have to wait 3 PTO before you can also remove the path id/close the path entirely.

Or alternatively, you always have to 3 PTOs after the abandon farm before you remove any state including the last used CID but that also means that the retired CID frames doesn't have any impact. And it also means whenever you received a retired CID frame during an active path use, you also need to wait 3 PTOs before you remove it.

Reply:
The question about "what to do with packets where DCID is retired" is not specific to multipath. The same "out of order delivery issues" could happen with QUIC as well. After a CID is retired, we expect the issuer of the CID to "forget about it". If that CID is forgotten, the load balancer or the QUIC server will not be able to associate it with a QUIC connection context. The QUIC connection will never have to process these packets.

The QUIC server will assume that the client is trying to use a connection that does not exist anymore and could issue a "stateless reset" packet, but since the CID has been retired the token in the stateless reset will not match any valid CID for the connection and will have no effect.

Which is a long winded way to say that packets arriving after the DCID has been retired will be automatically ignored. There is no need to specify a processing for them, because they will never be processed at all.